### ISSUE TITLE
Possible assertion failure in ldelf with 64-bit TAs

### ISSUE THREAD

**Issue by jforissier** [OP-TEE/optee_os#3118]
**TL;DR**

OP-TEE 3.6.0-rc1 might fail to load some 64-bit TAs and report an assertion
failure in `ldelf`. The problematic TAs are when the linker puts `__ta_entry()`
in first position in `.text`, immediately after `.ta_head`. Unlikely, but can
happen.
The fix in `ldelf` is not obvious (to me). In the meantime, there is an easy
workaround with the TA linker script. Any problematic TA can be re-linked
and the issue will go away.

**Problem description**

After rebasing my Clang support RFC (PR #2944) onto current master, which has
the new user-space ELF loader (`ldelf`), I met an assertion failure in `ldelf`.
The TEE would refuse to load the TA:
```
E/LD:  assertion 'head->depr_entry != UINT64_MAX' failed at ldelf/ta_elf.c:889 in ta_elf_load_main()
```
It turns out my Clang support code triggers a bug that is already present in
the master branch.

**Analysis**

The problem is that `elf->e_entry == 0x20` (`== sizeof(*head)`), yet the TA is *not*
a legacy one (i.e., it is built with the "new" linker script that comes with the
`ldelf` patches). Not being legacy, it has `head->depr_entry == UINT64_MAX` which
causes the assertion failure.

The root cause seems to be this assumption:
```C
        /*
         * Legacy TAs doesn't set entry point, instead it's set in ta_head.
         * If entry point isn't set explicitly, [it is] set to the start of the
         * first executable section by the linker. Since ta_head also
         * always comes first in legacy TA it means that the entry point
         * will be set to 0x20.
         *
         * NB, everything before the commit a73b5878c89d ("Replace
         * ta_head.entry with elf entry") is considered legacy TAs for
         * ldelf.
         */
        if (elf->e_entry == sizeof(*head))
                elf->is_legacy = true;
```
I think this is backwards: legacy implies `e_entry == sizeof(*head)`, but
*not* the other way around. Indeed:

- Legacy TA files have the following layout:
```
  ELF header
    e_entry = 0x20 (set to the VA of .text by the linker because not set explicitly)
  First loadable segment
    .ta_head (VA = 0x0, size = 0x20 bytes)
        entry = <VA of __utee_entry or __ta_entry depending on version>
    .text (VA = 0x20)
        ... code ...
```
- Non-legacy TAs are as follows:
```
  ELF header
    e_entry = <VA of __ta_entry> (set by ENTRY(__ta_entry) in the new linker script)
  First loadable segment
    .ta_head (VA = 0x0, size = 0x20 bytes)
        depr_entry = UINT64_MAX
    .text (VA = 0x20)
        ... code ...
```
So, if `__ta_entry()` happens to be located at the very begining of `.text` in a non-
legacy TA, we will have `e_entry == 0x20` and `depr_entry == UINT64_MAX` hence the
assert.

**Reproduce on 3.6.0-rc1**

The following patch will reproduce the issue on the master branch with any TA
(32 or 64 bits). It places `__ta_entry()` just after the TA header, at address
`0x20`. In order to do that:
1. `__section(".text.__ta_entry")` and the corresponding entry in `ta.ld.S` will
place the function just after the TA header
2. In 32-bit mode, `__attribute__((target("arm")))` forces arm instructions
instead of thumb, because with thumb the entry address is offset by one
(`0x20` -> `0x21`) to indicate thumb mode.

With 32-bit TAs and without the reproducer patch, the bug cannot happen on
master because `__ta_entry()` is always compiled in thumb mode (`-mthumb` added
to `ta_arm32-platform-cflags` by `core/arch/arm/arm.mk`). Therefore the entry point
is always an odd value, so cannot be `0x20`.

However, I think 64-bit TAs can trigger the bug.

Reproducer patch
```diff
diff --git a/ta/arch/arm/ta.ld.S b/ta/arch/arm/ta.ld.S
index 019b6e7a66..449acbf6cc 100644
--- a/ta/arch/arm/ta.ld.S
+++ b/ta/arch/arm/ta.ld.S
@@ -23,6 +23,7 @@ SECTIONS {
 	.ta_head : {*(.ta_head)}
 	.text : {
 		__text_start = .;
+		*(.text.__ta_entry)
 		*(.text .text.*)
 		*(.stub)
 		*(.glue_7)
diff --git a/ta/arch/arm/user_ta_header.c b/ta/arch/arm/user_ta_header.c
index 558f505838..469aa82324 100644
--- a/ta/arch/arm/user_ta_header.c
+++ b/ta/arch/arm/user_ta_header.c
@@ -31,7 +31,12 @@ TEE_Result __utee_entry(unsigned long func, unsigned long session_id,
 void __noreturn __ta_entry(unsigned long func, unsigned long session_id,
 			   struct utee_params *up, unsigned long cmd_id);
 
-void __noreturn __ta_entry(unsigned long func, unsigned long session_id,
+void __noreturn
+	__section(".text.__ta_entry")
+#ifdef ARM32
+	__attribute__((target("arm")))
+#endif
+	__ta_entry(unsigned long func, unsigned long session_id,
 			   struct utee_params *up, unsigned long cmd_id)
 {
 	TEE_Result res = TEE_SUCCESS;

```

**Fix**

In order to reliably determine legacy vs non-legacy, I think we should rely
on the header only (`depr_entry`). Unfortunately, the header (and all loadable
segments) are mapped by `map_segments()` which needs `is_legacy` to select between
`populate_segments()` and `populate_segments_legacy()`. So there is a circular
dependency. Perhaps the first page of the first loadable segment could be mapped
temporarily?

**Workaround**

There is an easy workaround: we can make sure that a non-legacy TA will never
have its entry point set to `0x20` (`== sizeof(*ta_head)`) by introducing some
padding after ta_head in the linker script.

Workaround patch
```diff
diff --git a/ta/arch/arm/ta.ld.S b/ta/arch/arm/ta.ld.S
index bb9615ab32..492010dc46 100644
--- a/ta/arch/arm/ta.ld.S
+++ b/ta/arch/arm/ta.ld.S
@@ -21,6 +21,7 @@ OUTPUT_ARCH(aarch64)
 
 SECTIONS {
 	.ta_head : {*(.ta_head)}
+	. += 1; /* Workaround ldelf: 'head->depr_entry != UINT64_MAX' failed */
 	.text : {
 		__text_start = .;
 		*(.text .text.*)
```

**Comment by jbech-linaro** [comment]
So basically just moving forward here with no change for the release seems to be a bad decision. That gives three alternatives as I see it:
1) Revert ldelf patches.
2) Use your work around.
3) Push the release date.

Least invasive for now seems to be "2", i.e., use your workaround. With that we have something that is acceptable and works, but might not be the best solution. But with that applied we can go head with the release and then after the release we can try to come up with a proper solution.

**Comment by jforissier** [comment]
> So basically just moving forward here with no change for the release seems to be a bad decision.

Well, not necessarily. Given that the workaround ("2") is in the TA linker script and not in OP-TEE itself, it means that if someone ever encounters an issue with a TA he/she can apply the workaround locally. The TA will work, and will still work with future versions of OP-TEE (with the proper fix).

So TBH I'm not sure which option I prefer, "2" or do nothing.

**Comment by jbech-linaro** [comment]
We have ~10-15 T-B tags for different platforms and builds, so it's seems unlikely that it'll show up.

**Comment by jforissier** [comment]
@jbech-linaro exactly.

**Comment by jbech-linaro** [comment]
Ok, let's move on with the release, but we keep this particular issue open and try to come up with something asap.

**Comment by jenswi-linaro** [comment]
Since we preferably need to know if it's a legacy TA or not before mapping ta_head we need a way to tell that from the ELF header. How about setting the `EI_OSABI` byte in the e_ident array to something specific? Like 255 which seems to be known as `ELFOSABI_STANDALONE`. We should probably set `EI_ABIVERSION` to `1` also.

**Comment by jforissier** [comment]
@jenswi-linaro using the `EI_OSABI` byte would be convenient, but I think it would be abusing the purpose a bit. Indeed, the ELF spec says:

> Some fields in other ELF structures have flags and values that have operating system and/or ABI specific meanings; the interpretation of those fields is determined by the value of this byte. If the object file does not use any extensions, it is recommended that this byte be set to 0."

We're not using any extensions, so I think we should rather fix our ELF parser ;-)

**Comment by jenswi-linaro** [comment]
Another option is to assume that a TA isn't a legacy TA and if it later turns out that it is we're just restarting the loading. It will mean that legacy TAs will be loaded twice so they will take a bit longer to load.

**Comment by jenswi-linaro** [comment]
https://github.com/OP-TEE/optee_os/pull/3158 should take care of this

**Comment by jbech-linaro** [comment]
@jforissier , please close this ticket if #3158 indeed fixes this.

**Comment by jforissier** [comment]
Fix #3158 verified.
